### DISCUSSION TITLE
Exists() within an Or() executes twice for each matching OR?

### DISCUSSION THREAD

**Discussion by douglasg14b** [NRules/NRules#337] [Q&A]
```cs
public class CustomerMessageAfterAssignmentRule : Rule
{
    public override void Define()
    {
        Regex customerFallback = new Regex("^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$");
        ExpertFirstAssignmentFact expertAssignment = null!;

        When()
            .Match(() => expertAssignment)
            .Or(or => or
                .Exists<CustomerMessageFact>(message => message.Timestamp >= expertAssignment.Timestamp)
                .Exists<UnknownMessageFact>(
                    message => message.Timestamp >= expertAssignment.Timestamp,
                    message => customerFallback.IsMatch(message.Message.Author!)
                )
            );

        Then()
            .Yield(_ => new CustomerMessageAfterAssignmentFact(expertAssignment));
    }
}
```

This will actually activate twice if either both OR conditions are matched.

Shouldn't this only execute once, breaking off early in the same way an expression would if it contained several `or` conditions?

**Comment by snikolayev** [comment]
@douglasg14b yes, this is expected. OR conditions are actually transformed into 2 rules behind the scenes, and if both rules match, then both will fire. 
If you can avoid OR conditions in your logic, I'd recommend you do so.

**Reply by douglasg14b** [reply]
Oh gotcha, hm, that's tough. 

Do you have recommended ways to approach logical `OR` conditions where circuit breaking semantics are desired? Or is that something that doesn't have explicit support?

**Bonus**: If that is something without explicit support, what do you consider the largest challenges if support was considered? Would this require deep knowledge of your implementation?  _(I'm trying to judge how difficult it's going to be to contribute to your lib)_

I change it to be:

```cs
        When()
            .Match(() => expertAssignment)
            .Or(or => or
                .Exists<CustomerMessageFact>(message => message.Timestamp >= expertAssignment.Timestamp)
                .And(and => and
                    .Not<CustomerMessageFact>(message => message.Timestamp >= expertAssignment.Timestamp)
                    .Exists<UnknownMessageFact>(
                        message => message.Timestamp >= expertAssignment.Timestamp,
                        message => customerFallback.IsMatch(message.Message.Author!)
                    )
                )
            );
```

However, that only gets worse if there are many fallbacks (I have a couple of these). Gotta love dirty data
